feat(api-service): api for env level change fixes NV-6155

scopsy opened the pull request:
### What changed? Why was the change needed?

<!-- Also include any relevant links, such as Linear tickets, Slack discussions, or design documents. -->

### Screenshots

<!-- If the changes are visual, include screenshots or screencasts. -->

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Related enterprise PR

<!-- A link to a dependent pull request  -->

### Special notes for your reviewer

<!-- Specific instructions or considerations you want to highlight for the reviewer. -->

</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Introduced API endpoints to compare ("diff") and publish resources (workflows) between environments, including dry-run support.
  * Added detailed reporting on changes (added, modified, deleted, unchanged) for workflows and their steps between environments.
  * Enabled transactional synchronization to ensure safe, atomic publishing of workflows.
  * Provided comprehensive summaries and structured results for both diff and publish operations.

* **Bug Fixes**
  * Ensured payload schema and validation settings are preserved when syncing workflows across environments.

* **Tests**
  * Added end-to-end tests for environment diff and publish endpoints, covering validation, resource changes, and response structure.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

scopsy committed:
fix: wip

scopsy committed:
fix: wip

scopsy committed:
fix: fix

scopsy committed:
l

scopsy committed:
fix: it

scopsy committed:
diz: q

scopsy committed:
fix

scopsy committed:
feat: add test

scopsy committed:
Merge branch 'next' into nv-6155-api-for-env-level-change

scopsy committed:
Update pnpm-lock.yaml

scopsy committed:
Replace WorkflowOriginEnum with ResourceOriginEnum

Updated imports and usage of WorkflowOriginEnum to ResourceOriginEnum in environment and workflow sync logic to align with changes in the shared module. This ensures consistency and compatibility with the updated enum naming.

scopsy committed:
Add step-level diff support to environment diff

Enhanced the environment diff logic to include step-level changes (added, modified, deleted, moved) in workflows. Updated DTOs, types, and sync strategies to support detailed step diffs and extended summary statistics. This provides more granular insight into workflow differences between environments.

scopsy committed:
Refactor diff to return per-entity results with metadata

The diff logic now returns an array of results, each representing a single entity (workflow) with its own entityId, entityName, diffs, and summary. DTOs, types, and tests were updated to reflect this structure, improving clarity and granularity of environment diff results.

scopsy committed:
Simplify diff summary by merging step and workflow counts

Removed separate step-level change counters from diff summary DTOs, types, and aggregation logic. Now, added, modified, deleted, and unchanged counts include both workflows and steps, streamlining the summary structure and reducing redundancy.

scopsy committed:
Update workflow-sync.strategy.ts

scopsy committed:
Support workflow deletion during environment sync

Added logic to detect and delete workflows from the target environment when they are removed from the source during environment sync. Updated DTOs and types to include the 'deleted' action, extended the workflow sync strategy to handle deletions, and added an e2e test to verify this behavior. Also exported DeleteWorkflowUseCase from the workflow module.

scopsy committed:
Merge branch 'next' into nv-6155-api-for-env-level-change

scopsy committed:
Include normalized step data in step diff changes

Adds the normalized step data to the 'changes' field when steps are added or deleted in the workflow sync strategy. This provides more detailed information about the step changes for downstream consumers.

scopsy committed:
Rename 'old' to 'previous' in diff-related code

Replaces all occurrences of the 'old' property with 'previous' in DTOs, types, and sync strategy logic for environment diffs. This change improves clarity and consistency in naming for previous values and indices throughout the codebase.

scopsy committed:
Remove duration and timing fields from sync results

Eliminated duration and timing-related fields from DTOs, types, and sync logic for environment publishing and workflow synchronization. This simplifies the API response and internal data structures by removing per-entity and total timing metrics, as they are no longer required.

scopsy committed:
Update publish-environment.dto.ts

scopsy committed:
Refactor publish summary property names for clarity

Renamed summary properties in the publish environment flow from totalEntities, totalSuccessful, totalFailed, and totalSkipped to entities, successful, failed, and skipped for improved clarity and consistency. Updated all related usages in types, use case logic, and e2e tests.

scopsy committed:
Remove includeInactive option from environment diff

Eliminated the includeInactive parameter from DTOs, commands, controller, and workflow sync strategy. All workflow diff and sync operations now consider only active workflows by default, simplifying the API and internal logic.

scopsy committed:
Remove includeInactive option from environment publishing

The includeInactive property and related logic have been removed from DTOs, controller, types, command, and use case for publishing environments. This simplifies the API and internal logic by no longer supporting the inclusion of inactive workflows during environment sync.

scopsy committed:
Remove skipExisting option from publish environment flow

Eliminated the skipExisting property from DTOs, command, types, controller, and use case for publishing environments. This streamlines the API and internal logic by no longer supporting the option to skip existing workflows during environment publishing.

scopsy committed:
Remove batchSize option from publish environment flow

Eliminated the batchSize property from the PublishEnvironment DTO, command, and controller. The batch size is now hardcoded to 100 in the use case, simplifying the API and internal logic.

scopsy committed:
Remove unused sync result helpers from base strategy

Deleted the createSyncResult and createDiffResult helper methods from BaseSyncStrategy as they are no longer used. Moved a private createSyncResult implementation to WorkflowSyncStrategy to keep logic local to its usage. Also removed the unused executeWithoutTransaction method from TransactionalSyncService.

scopsy committed:
Update workflow-sync.strategy.ts

scopsy committed:
Refactor workflow sync strategy with modular operations

Refactored the workflow sync strategy to delegate sync and diff logic to dedicated operation classes, improving modularity and maintainability. Introduced builders, comparators, normalizers, and constants for workflow sync, and registered these components in the module. This change separates concerns and prepares the codebase for easier testing and extension.

scopsy committed:
Add userContext to workflow sync diff operations

Refactored workflow sync strategy and related operations to accept a userContext parameter for diff and comparison methods. This change removes hardcoded user data and allows passing dynamic user context, improving flexibility and security in workflow diff and sync processes.

scopsy committed:
Update workflow-sync.constants.ts

scopsy committed:
Refactor sync types to use strong typing and enums

Replaces string literals for entity types with EntityTypeEnum and adds strong typing for user context (UserSessionData) and workflow entities (NotificationTemplateEntity) across sync strategies, comparators, and operations. Improves type safety and consistency in environment sync and diff logic.

scopsy committed:
Use EntityTypeEnum.WORKFLOW for entityType

Replaces hardcoded 'workflow' string with EntityTypeEnum.WORKFLOW in diff result builder to improve type safety and consistency.

scopsy committed:
Refactor entity to resource terminology in sync logic

Replaces all references to 'entity' with 'resource' across DTOs, types, use cases, builders, and tests for environments-v2 sync and diff logic. This improves clarity and consistency in naming, reflecting that workflows and steps are treated as resources. Updates all affected interfaces, classes, and test assertions accordingly.

scopsy committed:
Use workflow identifier instead of _id in sync and diff ops

Replaced usage of workflow._id with a new getWorkflowIdentifier method in both WorkflowDiffOperation and WorkflowSyncOperation. This ensures that workflow operations consistently use the workflow's trigger identifier as the resource ID, improving consistency and reliability when referencing workflows.

scopsy committed:
Refactor diff DTOs and types for source/target clarity

Updated diff-related DTOs, types, and builders to distinguish between source and target resource IDs and names, improving clarity in diff results. Adjusted changes structure to group previous and new states, and updated tests and workflow comparator logic accordingly. This refactor enhances the expressiveness and accuracy of environment diff and sync operations.

scopsy committed:
Update environments-v2-diff.e2e.ts

scopsy committed:
Refactor workflow diff to unify step actions and types

Consolidates step-related diff actions into standard actions ('added', 'modified', 'deleted', 'moved') and removes step-specific fields (workflowId, workflowName) from diff DTOs and types. Updates all affected interfaces, enums, and builder/comparator logic to use the new unified action types and nullable change objects. This simplifies the diff model and improves consistency across resource and step diffs.

scopsy committed:
Rename 'diffs' and 'changes' fields for consistency

Standardizes the naming of diff result fields across DTOs, types, builders, comparators, and tests. The 'diffs' field is renamed to 'changes' in result objects, and the 'changes' field within individual diffs is renamed to 'diffs'. Updates all relevant usages and tests to match the new naming convention for improved clarity and consistency.

scopsy committed:
Merge branch 'next' into nv-6155-api-for-env-level-change

scopsy committed:
Refactor workflow sync builders for resource type support

Updated SyncResultBuilder and DiffResultBuilder to accept a resource type, enabling support for resources beyond workflows. Adjusted dependency injection in the module and operations to provide builders with the correct resource type. Deprecated workflow-specific builder methods in favor of generic resource methods, maintaining backward compatibility. Also removed an unused timing utility from BaseSyncStrategy.

scopsy committed:
Merge branch 'nv-6155-api-for-env-level-change' of https://github.com/novuhq/novu into nv-6155-api-for-env-level-change

scopsy committed:
Refactor workflow normalization and update permissions

Changed the diffEnvironment endpoint to require ENVIRONMENT_READ instead of ENVIRONMENT_WRITE. Refactored workflow normalization logic to simplify and align step and workflow normalization, removing unnecessary fields and methods. Updated type definitions for normalized workflow and step to use Pick and match the new normalization logic.

scopsy committed:
Refactor workflow sync to use repository service

Introduces WorkflowRepositoryService to encapsulate workflow fetching and mapping logic, removing direct repository access from operations. Refactors WorkflowDiffOperation and WorkflowSyncOperation to use the new service, simplifies builder usage, and updates normalization and type definitions for workflows and steps. Removes unused builder providers and cleans up related interfaces.

scopsy committed:
Update workflow diff permissions and refactor types

Changed the diffEnvironment endpoint to require ENVIRONMENT_WRITE permission instead of ENVIRONMENT_READ. Cleaned up imports and comments in environments.module.ts. Refactored workflow-sync.types.ts to improve type safety by using Partial<INormalizedWorkflow> in IWorkflowComparison and removed unused imports.

scopsy committed:
Update environments.controller.ts

scopsy committed:
Update tests to use new summary fields in publish API

Replaces references to 'totalSuccessful' and 'totalSkipped' with 'successful' and 'skipped' in environment publish E2E tests to match updated API response fields. Also changes test setup from 'before' to 'beforeEach' to ensure isolation between tests.

scopsy committed:
Add updatedBy user tracking to workflows

Introduces the updatedBy field to workflow entities, DTOs, and database models to track the user who last updated a workflow. Updates mappers, use cases, commands, and repository logic to support population and serialization of the updatedBy user. This enables API consumers to see which user last modified a workflow.

scopsy committed:
Merge branch 'next' into nv-6155-api-for-env-level-change

scopsy committed:
Add updatedBy user info to environment diff results

Introduces user information fields (sourceResourceUpdatedBy, targetResourceUpdatedBy) to diff DTOs, types, and diff result builder for workflows and steps. Updates the workflow diff operation to extract and include the user who last updated each resource in the diff response. Adds and extends e2e tests to verify the presence and correctness of updatedBy information in diff results.

scopsy committed:
Add updatedAt fields to workflow diff results

Introduces sourceResourceUpdatedAt and targetResourceUpdatedAt fields to workflow diff DTOs, types, and builder logic. Updates the diff operation to extract and include updatedAt timestamps for both source and target resources. Enhances e2e tests to verify presence and correctness of updatedAt fields. Also refactors workflow repository to use a new findPublishable method for fetching workflows.

scopsy committed:
Refactor resource diff structure to use nested objects

Replaces flat resource ID, name, updatedBy, and updatedAt fields in diff DTOs and types with nested ResourceInfo objects for both source and target resources. Updates all usages, builders, comparators, and tests to use the new structure, improving consistency and extensibility of resource metadata handling in environment diffs.

scopsy committed:
Add environment type support and defaulting logic

Introduces the EnvironmentTypeEnum to environments, adds type field to environment entities, DTOs, and schemas, and ensures environments are assigned a type (dev or prod) based on their name or explicit input. Updates API, validation, and dashboard dropdown to support and display environment types, and provides backward compatibility for legacy environments without a type.

scopsy committed:
Merge branch 'next' into nv-6155-api-for-env-level-change

scopsy committed:
Update environment.schema.ts

scopsy committed:
Remove payload schema fields from workflow sync DTOs

Eliminated the inclusion of 'payloadSchema' and 'validatePayload' fields when mapping workflow data for environment synchronization. This streamlines the DTOs and prevents unnecessary data from being transferred.

scopsy committed:
Add MongoDB session support for transactional sync

Introduces optional ClientSession parameters to repository methods and command/usecase layers to enable transactional operations across workflow and environment sync processes. This change ensures atomicity and consistency during multi-step sync and upsert operations by passing the session context through the call stack.

scopsy committed:
Merge branch 'next' into nv-6155-api-for-env-level-change

scopsy committed:
Add feature flag for new environment type mechanism

Introduces the IS_NEW_CHANGE_MECHANISM_ENABLED feature flag to control how environment types are assigned when creating environments. Updates the environment creation use case and tests to respect this flag, defaulting all new environments to DEV type when the flag is disabled. Also refactors imports and updates feature flag enums.

scopsy committed:
Refactor environment validation into dedicated service

Extracted environment validation logic from use cases into a new EnvironmentValidationService. Updated DiffEnvironmentUseCase and PublishEnvironmentUseCase to use the new service, improving code reuse and maintainability.

scopsy committed:
Support non-replica set MongoDB by allowing null sessions

Refactored all repository and use case methods to accept `ClientSession | null` for session parameters, enabling operations to proceed without transactions when MongoDB is not running in replica set mode. Updated transaction handling to gracefully fallback to non-transactional execution if transactions are not supported. Adjusted command and use case signatures, and improved logging for transactional operations.

scopsy committed:
Exclude session property from command serialization

Added the @Exclude() decorator to the session property in workflow and message template command classes to prevent the session from being serialized in API responses. This change improves security and ensures sensitive session data is not exposed.

scopsy committed:
Update get-my-environments.usecase.ts

scopsy committed:
Add userId to GetMyEnvironmentsCommand

The userId is now passed from the controller to the GetMyEnvironmentsCommand. This allows use cases to access the user's ID when retrieving environments.

scopsy committed:
Improve workflow comparison and command serialization

Refactored environment diff E2E tests to use beforeEach for better isolation. Added null checks for source and target workflows in WorkflowComparator to prevent errors. Marked the session property with @Exclude in UpdateWorkflowCommand to avoid serializing it in responses.

scopsy committed:
Add updatedBy and type fields to workflow and environment DTOs

Introduces the 'updatedBy' field to WorkflowResponseDto and WorkflowListResponseDto, allowing tracking of the user who last updated a workflow. Adds the 'type' field to EnvironmentResponseDto as an optional property. Also updates related schemas and types in the internal SDK, and exposes the 'slug' field in LayoutResponseDto. Removes @ApiExcludeController from LayoutsController and improves OpenAPI property annotations for user DTOs.

scopsy committed:
Merge branch 'next' into nv-6155-api-for-env-level-change

scopsy committed:
Exclude LayoutsController from Swagger docs

Added the @ApiExcludeController() decorator to prevent the LayoutsController from appearing in the generated Swagger API documentation.

djabarovgeorge commented on apps/api/src/app/environments-v1/dtos/environment-response.dto.ts:
```diff
@@ -30,6 +31,14 @@ export class EnvironmentResponseDto {
+    enum: EnvironmentTypeEnum,
+    description: 'Type of the environment',
+    example: EnvironmentTypeEnum.PROD,
+    nullable: true,
```
```suggestion
]```

djabarovgeorge commented on apps/api/src/app/environments-v1/dtos/environment-response.dto.ts:
```diff
@@ -30,6 +31,14 @@ export class EnvironmentResponseDto {
   })
   identifier: string;
 
+  @ApiPropertyOptional({
+    enum: EnvironmentTypeEnum,
+    description: 'Type of the environment',
+    example: EnvironmentTypeEnum.PROD,
+    nullable: true,
+  })
+  type: EnvironmentTypeEnum;
```
```suggestion
  @ApiPropertyOptional({
    enum: EnvironmentTypeEnum,
    description: 'Type of the environment',
    example: EnvironmentTypeEnum.PROD,
  })
  type: EnvironmentTypeEnum;
```

scopsy replied:
@djabarovgeorge from what i've i've read in the docs, nullable !== required, where apiproprty optional only handeled the undefined state, nullable is different. and is explicitly added. 

<img width="2216" height="382" alt="Image" src="https://github.com/user-attachments/assets/24004585-d028-4fca-b361-c18254812fea" />

djabarovgeorge commented on apps/api/src/app/environments-v2/environments.controller.ts:
```diff
@@ -35,4 +58,45 @@ export class EnvironmentsController {
     );
   }
+
+  @Post('/publish')
```
I don't have any strong opinions, but I wonder if we should make those endpoints more restful so the route will be `/:environmentId/publish` where `environmentId` is the targetEnvironmentId

scopsy replied:
Was also thinking about it, so basically you suggest that source is always Dev by default and could be overriden with another one?

djabarovgeorge commented on apps/api/src/app/environments-v2/services/transactional-sync.service.ts:
were there any reason not making this usecase a function?
i am asking because i see that it is used only once

scopsy committed:
Merge branch 'next' into nv-6155-api-for-env-level-change

scopsy committed:
Remove TransactionalSyncService and refactor usage

Deleted TransactionalSyncService and replaced its usage with a private method in PublishEnvironmentUseCase. Updated module and service exports to remove TransactionalSyncService references, simplifying dependency management and reducing indirection.

scopsy committed:
Refactor environment diff and publish endpoints

Changed diff and publish API endpoints to use '/v2/environments/:targetEnvironmentId/diff' and '/v2/environments/:targetEnvironmentId/publish' instead of passing targetEnvironmentId in the request body. Updated DTOs, commands, use cases, controller, validation service, and tests to support this change. Also, sourceEnvironmentId is now optional and defaults to the development environment if not provided.

scopsy committed:
Refactor environment e2e tests and update SDK formatting

Improves e2e tests for environment diff and publish flows by clarifying workflow creation, adding skipped tests for known issues, and ensuring default source environment logic is tested. Refactors SDK class to use consistent import formatting and improves method argument formatting for readability.

scopsy committed:
Refactor workflow creation in e2e tests to use SDK

Updated environment diff and publish e2e tests to create workflows using the novuClient SDK instead of the previous createWorkflow function. Also added a short delay after workflow creation to ensure workflows are fully created before proceeding with tests. Removed skipped tests related to workflow creation issues.

scopsy committed:
Add session support to workflow and template queries

Introduces optional ClientSession parameters to workflow and notification template commands, usecases, and repository methods. This enables transactional operations for workflow creation, patching, upserting, and preference management, improving consistency and reliability in multi-step database operations.

scopsy committed:
Update publish test to reflect skipped workflow

Removed unnecessary comments and adjusted assertions in the environment publish E2E test. The test now expects the workflow to be skipped (successful: 0, skipped: 1) when publishing without a sourceEnvironmentId, reflecting updated behavior.

scopsy committed:
Update .source
